n3c_owl_ingest/n3c_owl_ingest.py

```diff
@@ -1,11 +1,74 @@
 """N3C OMOP to OWL"""
+import os
+from pathlib import Path
+from typing import Dict
+
+import pandas as pd
+
+CURIE = str
+SRC_DIR = Path(os.path.dirname(os.path.abspath(__file__)))
+PROJECT_DIR = SRC_DIR.parent
+TERMHUB_CSETS_DIR = PROJECT_DIR / 'termhub_csets'
+DATASETS_DIR = TERMHUB_CSETS_DIR / 'datasets' / 'prepped_files'
+CONCEPT_CSV = DATASETS_DIR / 'concept.csv'
+CONCEPT_RELATIONSHIP_CSV = DATASETS_DIR / 'concept_relationship_subsumes_only.csv'
+# - doesn't seem like concept_ancestor is necessary
+# ancestor_concept_id,descendant_concept_id,min_levels_of_separation,max_levels_of_separation
+# CONCEPT_ANCESTOR_CSV = DATASETS_DIR / 'concept_ancestor.csv'
 
 
 def run_ingest():
     """Run the ingest"""
-    print("Hello World")
+    # concept: header & example row
+    # concept_id,concept_name,domain_id,vocabulary_id,concept_class_id,standard_concept,concept_code,valid_start_date,valid_end_date,invalid_reason
+    # 22945,Horizontal overbite,Condition,SNOMED,Clinical Finding,S,70305005,2002-01-31,2099-12-31,
+    concept_df = pd.read_csv(CONCEPT_CSV)
+
+    # todo: use concept_relationship after constructing omop_d &/or native_d
+    # concept: header & example row
+    # concept_id_1,concept_id_2,relationship_id,valid_start_date,valid_end_date,invalid_reason
+    # 260759,22945,Subsumes,2011-07-31,2099-12-31,
+    concept_relationship_df = pd.read_csv(CONCEPT_RELATIONSHIP_CSV)
+
+    omop_d: Dict[CURIE, Dict[str, str]] = {}
```

Comment:
@Sigfried This is my primary question right now, is which of these hierarchy types to use, if not both. These dictionaries would get converted into OWL.
```python
omop_d: Dict[CURIE, Dict[str, str]] = {}
native_d: Dict[CURIE, Dict[str, str]] = {}
```
(a) `omop_d`: This hierarchy would be based on OMOP IDs, which I believe is what we're currently doing.
(b) `native_d`: I feel like we'll eventually run into a use case for this. This dictionary would store all terms, the keys of which would be their native CURIE, e.g. `SNOMED:123`. So it would get serialized into multiple hierarchies.

Possibly just (a) for now might be all we need. Do you agree?

Reply:
Thanks @joeflack4. Yes. (a) for now. I'm imagining you do the concept table and then you link the concepts (using concept_id) based on both the concept_relationship and concept_ancestor tables. I don't know if links can have properties.
<img width="527" alt="image" src="https://user-images.githubusercontent.com/1586931/230509002-a225eb6f-687a-44d7-aa41-28eb4d0420da.png">
For an initial pass, the relationship links can be limited to ones where invalid_reason is null and only need to include the two concept_ids and the relationship_id (though, if it's just as easy, include all the fields in that table; we'll use them eventually), and the ancestor table just needs the two ids and the min_levels_of_separation (though we might use max_levels_of_separation eventually.)

Reply:
@Sigfried Glad you reviewed this!

## Whether to parse `concept_ancestor`, why, and ramifications
For `concept_ancestor`, this _may_ not provide us anything that we can use via OAK. Do / will we have functionality in TermHub we ask **(i)** "is 'x' a descendant of 'y'?" or **(ii)** "how many min/max levels of separation are there between 'x' and 'y'?". If so, then even if I am to get this information somehow into OWL (I'm not sure the best way to do it; I'd have to find some predicates / relationship types or make some custom ones like `OMOP:is_descendant`, etc. But even if I do that, OAK will not be hard-wired to utilize these predicates with a user-friendly Python API, though it likely has some way to access them.

On the other hand, I would imagine (though I'm not 100% sure) that, if I simply ignore `concept_ancestor` and only parse `concept_relationship`, OAK should be able to answer **'i'**; not sure about **'ii'**. This is the kind of thing I was getting at when I mentioned in our meeting with Chris that we might have some feature requests. If you'd like, I can ping him to see if this is currently possible, but we could also check the documentation / play with OAK first.

It sounds like your'e saying we do need `min_levels_of_separation`, though. Regarding that, see next section.

## Can links / relationships have properties
Yes, they can. It's more advanced OWL/RDF that I'm not so good at yet. I have to look into it; this will slow things down more than my initial time estimate. I'm more concerned though with how to query those relationship properties (using OAK).

---

Other that, I hare you on `invalid_reason` and parsing out all properties in the table.

Reply:
Ok. The main purpose of the concept_ancestor table is to quickly query for all descendants/ancestors of a concept. If OAK will be plenty fast without that, maybe we don't need it. We probably don't need number of levels. But we will want to get paths between concepts sometimes.